File: backend/pkg/core/profile/service.go
package profile

import (
	"context"
	"encoding/json"

	"backend/pkg/core"
	"backend/pkg/types"

	"github.com/google/uuid"
)

// Service provides business logic for profile operations
type Service struct {
	profileStore      ProfileStore
	siteSettingsStore SiteSettingsStore
	accountStore      AccountStore
	organizationStore OrganizationStore
}

// NewService creates a new profile service with the provided stores
func NewService(
	profileStore ProfileStore,
	siteSettingsStore SiteSettingsStore,
	accountStore AccountStore,
	organizationStore OrganizationStore,
) *Service {
	return &Service{
		profileStore:      profileStore,
		siteSettingsStore: siteSettingsStore,
		accountStore:      accountStore,
		organizationStore: organizationStore,
	}
}

// GetPublicProfile retrieves the public profile based on site settings
func (s *Service) GetPublicProfile(ctx context.Context) (*PublicProfileResponse, error) {
	pub, err := s.profileStore.GetPublicProfile(ctx)
	if err != nil {
		return nil, err
	}

	socialLinks := make(map[string]string)
	if pub.SocialLinks != nil {
		data, _ := json.Marshal(pub.SocialLinks)
		json.Unmarshal(data, &socialLinks)
	}

	return &PublicProfileResponse{
		Type:            pub.Type,
		Name:            pub.Name,
		Bio:             stringValue(pub.Bio),
		ImageURL:        stringValue(pub.ImageURL),
		EmailPublic:     stringValue(pub.EmailPublic),
		SocialLinks:     socialLinks,
		MetaDescription: stringValue(pub.MetaDescription),
		WebsiteURL:      pub.WebsiteURL,
	}, nil
}

// GetUserProfile returns the profile for a specific user
func (s *Service) GetUserProfile(ctx context.Context, accountID uuid.UUID) (*UserProfile, error) {
	account, err := s.accountStore.FindByID(ctx, accountID)
	if err != nil {
		return nil, err
	}

	socialLinks := parseSocialLinksFromMap(account.SocialLinks)

	return &UserProfile{
		ID:              account.ID,
		Name:            account.Name,
		Bio:             stringValue(account.Bio),
		ProfileImage:    stringValue(account.ProfileImage),
		EmailPublic:     stringValue(account.EmailPublic),
		SocialLinks:     socialLinks,
		MetaDescription: stringValue(account.MetaDescription),
		OrganizationID:  account.OrganizationID,
	}, nil
}

// UpdateUserProfile updates the profile fields for a user
func (s *Service) UpdateUserProfile(ctx context.Context, accountID uuid.UUID, req ProfileUpdateRequest) (*UserProfile, error) {
	account, err := s.accountStore.FindByID(ctx, accountID)
	if err != nil {
		return nil, err
	}

	// Apply updates
	if req.Name != nil {
		account.Name = *req.Name
	}
	if req.Bio != nil {
		account.Bio = req.Bio
	}
	if req.ProfileImage != nil {
		account.ProfileImage = req.ProfileImage
	}
	if req.EmailPublic != nil {
		account.EmailPublic = req.EmailPublic
	}
	if req.MetaDescription != nil {
		account.MetaDescription = req.MetaDescription
	}
	if req.SocialLinks != nil {
		// Convert map[string]string to map[string]interface{}
		socialLinksInterface := make(map[string]interface{})
		for k, v := range *req.SocialLinks {
			socialLinksInterface[k] = v
		}
		account.SocialLinks = socialLinksInterface
	}

	if err := s.accountStore.Update(ctx, account); err != nil {
		return nil, err
	}

	socialLinks := parseSocialLinksFromMap(account.SocialLinks)

	return &UserProfile{
		ID:              account.ID,
		Name:            account.Name,
		Bio:             stringValue(account.Bio),
		ProfileImage:    stringValue(account.ProfileImage),
		EmailPublic:     stringValue(account.EmailPublic),
		SocialLinks:     socialLinks,
		MetaDescription: stringValue(account.MetaDescription),
		OrganizationID:  account.OrganizationID,
	}, nil
}

// GetSiteSettings returns the current site settings
func (s *Service) GetSiteSettings(ctx context.Context) (*SiteSettingsResponse, error) {
	settings, err := s.siteSettingsStore.Get(ctx)
	if err != nil {
		if err == core.ErrNotFound {
			// Return defaults
			return &SiteSettingsResponse{
				PublicProfileType: "user",
			}, nil
		}
		return nil, err
	}

	return &SiteSettingsResponse{
		PublicProfileType:    settings.PublicProfileType,
		PublicUserID:         settings.PublicUserID,
		PublicOrganizationID: settings.PublicOrganizationID,
	}, nil
}

// UpdateSiteSettings updates the site settings
func (s *Service) UpdateSiteSettings(ctx context.Context, req SiteSettingsUpdateRequest) (*SiteSettingsResponse, error) {
	settings, err := s.siteSettingsStore.Get(ctx)
	if err != nil {
		if err == core.ErrNotFound {
			// Create default settings
			settings = &types.SiteSettings{ID: 1, PublicProfileType: "user"}
		} else {
			return nil, err
		}
	}

	if req.PublicProfileType != nil {
		if *req.PublicProfileType != "user" && *req.PublicProfileType != "organization" {
			return nil, core.ErrValidation
		}
		settings.PublicProfileType = *req.PublicProfileType
	}
	if req.PublicUserID != nil {
		// Verify user exists using account store
		_, err := s.accountStore.FindByID(ctx, *req.PublicUserID)
		if err != nil {
			if err == core.ErrNotFound {
				return nil, core.ErrNotFound
			}
			return nil, err
		}
		settings.PublicUserID = req.PublicUserID
	}
	if req.PublicOrganizationID != nil {
		// Verify organization exists using organization store
		_, err := s.organizationStore.FindByID(ctx, *req.PublicOrganizationID)
		if err != nil {
			if err == core.ErrNotFound {
				return nil, core.ErrNotFound
			}
			return nil, err
		}
		settings.PublicOrganizationID = req.PublicOrganizationID
	}

	if err := s.siteSettingsStore.Save(ctx, settings); err != nil {
		return nil, err
	}

	return &SiteSettingsResponse{
		PublicProfileType:    settings.PublicProfileType,
		PublicUserID:         settings.PublicUserID,
		PublicOrganizationID: settings.PublicOrganizationID,
	}, nil
}

// IsUserAdmin checks if a user has admin role
func (s *Service) IsUserAdmin(ctx context.Context, userID uuid.UUID) (bool, error) {
	return s.profileStore.IsUserAdmin(ctx, userID)
}

// Helper functions

func stringValue(s *string) string {
	if s == nil {
		return ""
	}
	return *s
}

func parseSocialLinksFromMap(data map[string]interface{}) map[string]string {
	result := make(map[string]string)
	for k, v := range data {
		if str, ok := v.(string); ok {
			result[k] = str
		}
	}
	return result
}

File: backend/src/core/profile/service.rs
use std::{collections::BTreeMap, sync::Arc};

use serde_json::Value;
use uuid::Uuid;

use crate::{core::organization::OrganizationRepository, error::AppError};

use super::{
    ProfileAccount, ProfileAccountRepository, ProfileRepository, ProfileUpdateRequest,
    PublicProfileResponse, SiteSettings, SiteSettingsRepository, SiteSettingsResponse,
    SiteSettingsUpdateRequest, UserProfileResponse,
};

#[derive(Clone)]
pub struct ProfileService {
    profiles: Arc<dyn ProfileRepository>,
    settings: Arc<dyn SiteSettingsRepository>,
    accounts: Arc<dyn ProfileAccountRepository>,
    organizations: Arc<dyn OrganizationRepository>,
}

impl ProfileService {
    pub fn new(
        profiles: Arc<dyn ProfileRepository>,
        settings: Arc<dyn SiteSettingsRepository>,
        accounts: Arc<dyn ProfileAccountRepository>,
        organizations: Arc<dyn OrganizationRepository>,
    ) -> Self {
        Self {
            profiles,
            settings,
            accounts,
            organizations,
        }
    }

    pub async fn get_public_profile(&self) -> Result<PublicProfileResponse, AppError> {
        let profile = self.profiles.get_public_profile().await?;
        Ok(PublicProfileResponse {
            profile_type: profile.profile_type,
            // The Go service never assigns the DTO's ID field, so its encoded
            // value is UUID's all-zero value even when a backing row exists.
            id: Uuid::nil(),
            name: profile.name,
            bio: profile.bio.unwrap_or_default(),
            image_url: profile.image_url.unwrap_or_default(),
            email_public: profile.email_public.unwrap_or_default(),
            social_links: string_values(profile.social_links),
            meta_description: profile.meta_description.unwrap_or_default(),
            website_url: profile.website_url,
        })
    }

    pub async fn get_user_profile(
        &self,
        account_id: Uuid,
    ) -> Result<UserProfileResponse, AppError> {
        self.accounts
            .find_profile_account(account_id)
            .await
            .map(account_response)
    }

    pub async fn update_user_profile(
        &self,
        account_id: Uuid,
        request: ProfileUpdateRequest,
    ) -> Result<UserProfileResponse, AppError> {
        let mut account = self.accounts.find_profile_account(account_id).await?;
        if let Some(name) = request.name {
            account.name = name;
        }
        if let Some(bio) = request.bio {
            account.bio = Some(bio);
        }
        if let Some(profile_image) = request.profile_image {
            account.profile_image = Some(profile_image);
        }
        if let Some(email_public) = request.email_public {
            account.email_public = Some(email_public);
        }
        if let Some(meta_description) = request.meta_description {
            account.meta_description = Some(meta_description);
        }
        if let Some(social_links) = request.social_links {
            account.social_links = Some(
                social_links
                    .into_iter()
                    .map(|(key, value)| (key, Value::String(value)))
                    .collect(),
            );
        }
        self.accounts.update_profile_account(&account).await?;
        Ok(account_response(account))
    }

    pub async fn get_site_settings(&self) -> Result<SiteSettingsResponse, AppError> {
        match self.settings.get().await {
            Ok(settings) => Ok(settings_response(settings)),
            Err(AppError::NotFound) => Ok(SiteSettingsResponse {
                public_profile_type: "user".to_owned(),
                public_user_id: None,
                public_organization_id: None,
            }),
            Err(error) => Err(error),
        }
    }

    pub async fn update_site_settings(
        &self,
        request: SiteSettingsUpdateRequest,
    ) -> Result<SiteSettingsResponse, AppError> {
        let mut settings = match self.settings.get().await {
            Ok(settings) => settings,
            Err(AppError::NotFound) => SiteSettings {
                id: 1,
                public_profile_type: "user".to_owned(),
                public_user_id: None,
                public_organization_id: None,
                created_at: None,
                updated_at: None,
            },
            Err(error) => return Err(error),
        };

        if let Some(profile_type) = request.public_profile_type {
            if profile_type != "user" && profile_type != "organization" {
                return Err(AppError::InvalidInput("validation failed".to_owned()));
            }
            settings.public_profile_type = profile_type;
        }
        if let Some(user_id) = request.public_user_id {
            self.accounts.find_profile_account(user_id).await?;
            settings.public_user_id = Some(user_id);
        }
        if let Some(organization_id) = request.public_organization_id {
            self.organizations.find_by_id(organization_id).await?;
            settings.public_organization_id = Some(organization_id);
        }
        self.settings.save(&mut settings).await?;
        Ok(settings_response(settings))
    }

    pub async fn is_user_admin(&self, user_id: Uuid) -> Result<bool, AppError> {
        self.profiles.is_user_admin(user_id).await
    }
}

fn account_response(account: ProfileAccount) -> UserProfileResponse {
    UserProfileResponse {
        id: account.id,
        name: account.name,
        bio: account.bio.unwrap_or_default(),
        profile_image: account.profile_image.unwrap_or_default(),
        email_public: account.email_public.unwrap_or_default(),
        social_links: string_values(account.social_links),
        meta_description: account.meta_description.unwrap_or_default(),
        organization_id: account.organization_id,
    }
}

fn settings_response(settings: SiteSettings) -> SiteSettingsResponse {
    SiteSettingsResponse {
        public_profile_type: settings.public_profile_type,
        public_user_id: settings.public_user_id,
        public_organization_id: settings.public_organization_id,
    }
}

fn string_values(values: Option<BTreeMap<String, Value>>) -> BTreeMap<String, String> {
    values
        .unwrap_or_default()
        .into_iter()
        .filter_map(|(key, value)| value.as_str().map(|value| (key, value.to_owned())))
        .collect()
}
